### PR TITLE
feat: Reuse Dashboard redux data in Explore

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
Currently when user open Explore, we always send a request to `/v1/explore` endpoint, which assembles Explore initial data - slice, form_data and datasource.
When user goes to Explore from Dashboard (by clicking "Edit chart" or char title), most of the data needed to assemble Explore should already be available in Redux store.
This PR implements using data from Dashboard Redux store to initialize Explore without calling `/v1/explore` endpoint.

If user opens Explore from Dashboard, check if slice, formData, datasourceId and datasourceType are available. If they are, fetch datasource metadata and hydrate Explore using slice, formData and datasource.
If they are not available, call `/v1/explore` and hydrate Explore.

We can't reuse datasource from Dashboard. Datasources metadata on Dashboard is trimmed so that datasource contains only columns and metrics used by charts on that dashboard. In Explore we need to be able to access all columns and metrics, which means we must fetch those from API.

### TESTING INSTRUCTIONS
1. Go to dashboard
2. Click "Edit chart"
3. Verify that Explore has opened correctly and that we did not call `/v1/explore/` endpoint
4. Enter Explore from any other entry point (for example Charts list or Welcome page)
5. Verify that we do call `/v1/explore`

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/explore/ExplorePage.tsx [hunk 1]
```diff
@@ -38,12 +45,58 @@ const fetchExploreData = () => {
...
+      !shouldUseDashboardData ||
+      !isDefined(slice) ||
+      !isDefined(formData) ||
+      !isDefined(datasourceId) ||
+      !isDefined(datasourceType)
+    ) {
+      return fetchExploreData();
+    }
+    return SupersetClient.get({
+      endpoint: `/datasource/get/${datasourceType}/${datasourceId}/`,
```
Comment (zhaoyongjie) [inline superset-frontend/src/explore/ExplorePage.tsx:75]: there is a new endpoint for the dataset GET method is `/api/v1/dataset/<datasource PK(a.k.a datasource id without type>`. I added it at [here](https://github.com/apache/superset/pull/20677/files#diff-cc681838b8b8b6cb3d1a5dbf02c118d5f3353db8023f49c55f7a673e879be030)

**Review by zhaoyongjie — COMMENTED** [review]
leave a minor suggestion.

File: superset-frontend/src/explore/ExplorePage.tsx (hunk 1 above)
Comment (kgabryje) [reply superset-frontend/src/explore/ExplorePage.tsx:75]: I actually used `GET /v1/dataset` endpoint at first, but then changed it to `GET /datasource/get/`. As far as I know, Explore now supports (or will soon support?) other sources than datasets, such as SQL queries. Tagging @hughhhh for transparency.
Also, the `/v1/explore/` endpoint uses `DatasourceDAO` to get datasource and the response from `/v1/dataset` endpoint is missing some fields compared to datasource from `/v1/explore`, which caused some components in Explore to crash

Comment (kgabryje) [reply superset-frontend/src/explore/ExplorePage.tsx:75]: I propose we stick to the legacy `/datasource/get` endpoint for now and create a new `/v1/datasource` API in the future and refactor

**Review by zhaoyongjie — APPROVED** [review]

**Review by michael-s-molina — APPROVED** [review]
LGTM

**Comment by kgabryje** [comment]
This PR will be re-opened when linking from Dashboard to Explore is refactored (coming in a few days)

**Pull request merged**